pkg/applicationserver/grpc_deviceregistry.go

```diff
@@ -34,13 +35,39 @@ func (r asEndDeviceRegistryServer) Get(ctx context.Context, req *ttnpb.GetEndDev
 	return r.registry.Get(ctx, req.EndDeviceIdentifiers, req.FieldMask.Paths)
 }
 
+var errInvalidFieldValue = errors.DefineInvalidArgument("field_value", "invalid value of field `{field}`")
+
 // Set implements ttnpb.AsEndDeviceRegistryServer.
 func (r asEndDeviceRegistryServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest) (*ttnpb.EndDevice, error) {
 	if err := rights.RequireApplication(ctx, req.EndDevice.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
 		return nil, err
 	}
-	return r.registry.Set(ctx, req.EndDevice.EndDeviceIdentifiers, req.FieldMask.Paths, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
-		return &req.EndDevice, req.FieldMask.Paths, nil
+
+	if ttnpb.HasAnyField(req.FieldMask.Paths, "ids.dev_addr") && req.EndDevice.DevAddr != nil && !req.EndDevice.DevAddr.IsZero() {
+		return nil, errInvalidFieldValue.WithAttributes("field", "ids.dev_addr")
+	}
+
+	gets := append(req.FieldMask.Paths[:0:0], req.FieldMask.Paths...)
```

Comment:
Why

Reply:
We need to ensure `req.FieldMask.Paths` is not mutated on https://github.com/TheThingsNetwork/lorawan-stack/pull/776/files#diff-9d4e6149ff655290e12bcfb9931c439aR53 and https://github.com/TheThingsNetwork/lorawan-stack/pull/776/files#diff-9d4e6149ff655290e12bcfb9931c439aR56 and the only way to guarantee this is by copying

Reply:
E.g. consider https://play.golang.org/p/NKxIKiAEttg
Assuming registry implementation(s) only `append` and `len(req.FieldMask.Paths) == cap(req.FieldMask.Paths)` this in not strictly necessary, but I think it's a good and quite cheap safety measure to ensure correctness.

Reply:
I see, and it is a good safety measure, but aren't we protecting it in the wrong place?

I think that the registry `Set()` should not modify the getter. When it does, i.e. by appending implicit fields, it should do that to a copy.